PYIC-6240: Add generic audit event in processAsync for F2f & DCMAW to use

AmritSidhu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Make the existing process-async-VC audit events generic so they can be used for both F2F and DCMAW (and any future async CRIs). Include an additional extension credentialIssuerId so the CRI can be distinguished.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-6240](https://govukverify.atlassian.net/browse/PYIC-6240)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed


[PYIC-6240]: https://govukverify.atlassian.net/browse/PYIC-6240?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-6240: Add generic audit event in processAsync for F2f & DCMAW to use

Make the existing process-async-VC audit events generic so they can be used for both F2F and DCMAW (and any future async CRIs). Include an additional extension credentialIssuerId so the CRI can be distinguished.

DanCorderIPV commented on lambdas/initialise-ipv-session/src/test/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandlerTest.java:
```diff
@@ -886,7 +886,8 @@ void shouldSendAuditEventForIpvInheritedIdentityVcReceived() throws Exception {
                             Period.between(LocalDate.parse(TestVc.DEFAULT_DOB), LocalDate.now())
-                                    .getYears());
+                                    .getYears(),
+                            "hmrcMigration");
```
Why are we adding credential_issuer_id to a non-async audit event?

AmritSidhu committed:
PYIC-6240: credential_issuer_id is only added to async event

This change makes credential_issuer_id an optional field and therefore reverts the change to add it to all declarations

DanCorderIPV commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -245,19 +254,28 @@ void sendIpvVcConsumedAuditEvent(
+                        configService.getParameter(ConfigurationVariable.COMPONENT_ID),
+                        auditEventUser,
+                        extension,
+                        getRestrictedAuditDataForF2F(vc));
```
Should this be `getExtensionsForAuditIncCriId()`?

DanCorderIPV commented on libs/audit-service/src/test/java/uk/gov/di/ipv/core/library/auditing/helpers/AuditExtensionsHelperTest.java:
```diff
@@ -58,6 +60,24 @@ void shouldGetIdentityCheckVerifiableCredentialExtensionsForAudit() throws Excep
+        assertEquals(2, evidence.get(0).getValidityScore());
+        assertEquals(4, evidence.get(0).getStrengthScore());
+        assertEquals("bcd2346", evidence.get(0).getTxn());
+        assertEquals(2, evidence.get(0).getCheckDetails().size());
```
These tests should check the CRI ID value?

DanCorderIPV commented on libs/audit-service/src/main/java/uk/gov/di/ipv/core/library/auditing/helpers/AuditExtensionsHelper.java:
```diff
@@ -46,6 +46,32 @@ public static AuditExtensionsVcEvidence getExtensionsForAudit(
+        return new AuditExtensionsVcEvidence(
+                issuer, null, isSuccessful, vot, isUkIssued, age, criId);
+    }
+
```
Could this be simplified to something like
```
    public static AuditExtensionsVcEvidence getExtensionsForAudit(
            VerifiableCredential vc, Boolean isSuccessful) throws UnrecognisedVotException {
        return getExtensionsForAudit(vc, isSuccessful, null);
    }

    public static AuditExtensionsVcEvidence getExtensionsForAuditWithCriId(
            VerifiableCredential vc, Boolean isSuccessful) throws UnrecognisedVotException {
        return getExtensionsForAudit(vc, isSuccessful, vc.getCri().getId());
    }


    private static AuditExtensionsVcEvidence getExtensionsForAudit(
            VerifiableCredential vc, Boolean isSuccessful, String cridId) throws UnrecognisedVotException {
        var issuer = vc.getClaimsSet().getIssuer();
        var vot = VcHelper.getVcVot(vc);
        var isUkIssued = VcHelper.checkIfDocUKIssuedForCredential(vc);
        var age = VcHelper.extractAgeFromCredential(vc);

        List<?> evidence = null;

        if (vc.getCredential() instanceof IdentityCheckCredential identityCheckCredential) {
            evidence = identityCheckCredential.getEvidence();
        }

        if (vc.getCredential() instanceof RiskAssessmentCredential riskAssessmentCredential) {
            evidence = riskAssessmentCredential.getEvidence();
        }

        return new AuditExtensionsVcEvidence(issuer, evidence, isSuccessful, vot, isUkIssued, age, cridId);
    }
```

AmritSidhu committed:
PYIC-6240: Simplify code and assert for criId

DanCorderIPV commented on lambdas/process-async-cri-credential/src/main/java/uk/gov/di/ipv/core/processasynccricredential/ProcessAsyncCriCredentialHandler.java:
```diff
@@ -231,33 +232,51 @@ private void sendIpvVcReceivedAuditEvent(
+                        configService.getParameter(ConfigurationVariable.COMPONENT_ID),
+                        auditEventUser,
+                        getExtensionsForAuditWithCriId(verifiableCredential, isSuccessful),
+                        getRestrictedAuditDataForF2F(verifiableCredential));
```
This is still the F2F version?

AmritSidhu replied:
This is specific for restricted data (eg. names) which I believe is still required. I could generalise the method name or that can be done when we come back to refactor the f2f specific stuff out?

DanCorderIPV replied:
Oh, yes, I was looking at the wrong line. Would be good to generalise the method name now I think - assuming it won't take long

AmritSidhu committed:
PYIC-6240: Rename restricted data f2f class/method to be generic async
